File: wallet/src/de/schildbach/wallet/rates/ExchangeRatesDao.java
package de.schildbach.wallet.rates;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;

import org.dash.wallet.common.data.ExchangeRate;

import java.util.List;

import kotlinx.coroutines.flow.Flow;

/**
 * @author Samuel Barbosa
 */
@Dao
public interface ExchangeRatesDao {

    @Query("SELECT * FROM exchange_rates ORDER BY currencyCode")
    LiveData<List<ExchangeRate>> getAll();

    @Query("SELECT * FROM exchange_rates ORDER BY currencyCode")
    Flow<List<ExchangeRate>> observeAll();

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertAll(List<ExchangeRate> exchangeRates);

    @Query("SELECT * FROM exchange_rates WHERE currencyCode = :currencyCode LIMIT 1")
    LiveData<ExchangeRate> getRate(String currencyCode);

    @Query("SELECT * FROM exchange_rates WHERE currencyCode = :currencyCode LIMIT 1")
    Flow<ExchangeRate> observeRate(@NonNull String currencyCode);

    @Query("SELECT * FROM exchange_rates WHERE currencyCode = :currencyCode LIMIT 1")
    ExchangeRate getRateSync(String currencyCode);

    @Query("SELECT * FROM exchange_rates WHERE currencyCode LIKE :currencyCode || '%' ORDER BY currencyCode")
    LiveData<List<ExchangeRate>> searchRates(String currencyCode);

    @Query("SELECT count(*) FROM exchange_rates")
    int count();

    @Query("SELECT * FROM exchange_rates WHERE currencyCode = :currencyCode LIMIT 1")
    ExchangeRate getExchangeRateForCurrency(String currencyCode);
}

File: wallet/src/de/schildbach/wallet/database/dao/ExchangeRatesDao.kt
/*
 * Copyright 2023 Dash Core Group.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package de.schildbach.wallet.database.dao

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import kotlinx.coroutines.flow.Flow
import org.dash.wallet.common.data.entity.ExchangeRate

@Dao
interface ExchangeRatesDao {
    @Query("SELECT * FROM exchange_rates ORDER BY currencyCode")
    fun observeAll(): Flow<List<ExchangeRate>>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertAll(exchangeRates: List<ExchangeRate>)

    @Query("SELECT * FROM exchange_rates WHERE currencyCode = :currencyCode LIMIT 1")
    fun observeRate(currencyCode: String): Flow<ExchangeRate>

    @Query("SELECT * FROM exchange_rates WHERE currencyCode = :currencyCode LIMIT 1")
    fun getRateSync(currencyCode: String?): ExchangeRate?

    @Query("SELECT * FROM exchange_rates WHERE currencyCode = :currencyCode LIMIT 1")
    suspend fun getExchangeRateForCurrency(currencyCode: String?): ExchangeRate?

    @Query("SELECT count(*) FROM exchange_rates")
    suspend fun count(): Int

    @Query("DELETE FROM exchange_rates WHERE currencyCode IN (:currencyCodes)")
    suspend fun delete(currencyCodes: Collection<String>)

    @Query("SELECT * FROM exchange_rates ORDER BY currencyCode")
    suspend fun getAll(): List<ExchangeRate>
}
